source/snapshot/iterator.go

```diff
@@ -0,0 +1,182 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/conduitio/conduit-connector-postgres/source/position"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"gopkg.in/tomb.v2"
+)
+
+var ErrIteratorDone = errors.New("snapshot complete")
+
+type Config struct {
+	Position   sdk.Position
+	Tables     []string
+	TablesKeys map[string]string
+	Snapshot   string
+}
+
+type Iterator struct {
+	db      *pgxpool.Pool
+	t       *tomb.Tomb
+	workers []*FetcherWorker
+
+	conf Config
+
+	lastPosition position.Position
+
+	records chan sdk.Record
+}
+
+func NewIterator(ctx context.Context, conninfo string, c Config) (*Iterator, error) {
+	p, err := position.ParseSDKPosition(c.Position)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse position: %w", err)
+	}
+
+	if p.Snapshot == nil {
+		p.Snapshot = make(position.SnapshotPositions)
+	}
+
+	db, err := pgxpool.New(ctx, conninfo)
```

Comment:
Why are we creating a separate connection pool? 😢 We should let the caller give us a connection. My reasoning is that when doing long polling, we can get away with a single connection pool, we only need a separate connection if we do logical replication, in which case the caller (i.e. the combined iterator) should take care that it opens a separate connection and provides it here.

Reply:
Fair point, I did this to the fetchers but not so much in the iterator. Fixed in d4c8a39

The pool only keeps connection which were created via the `.Acquire()` method. Any other querier methods will do the acquire and release automagically. With Postgres only a single command per connection can be processed at the time. The fetchers each use a long running tx to pull the data, which then will automatically be released.